eduNEXT/xblock-mindmap
# feat: add support for grading mindmap components using submissions API

### Description
This PR adds support for grading mind-map components based on the submissions API. We draw inspiration from the [staff graded assignments xblock](https://github.com/mitodl/edx-sga/blob/master/edx_sga/sga.py) design.

To sup up the changes:
- First we configured the component as graded, adding `has_score` and the necessary configurations for the xblock like `weight` and `points`
- After we differentiated between `saving` and `submitting`, now the former saves the mindmap in the block state and the latter creates the submission using the API
- Then, we created the necessary handlers to manage grades: `enter_grade`, `remove_grade` including their helper methods
- We connected the handlers to the UI
- Finally, we created a way for easily testing the assignment of the grade by modifying the xblock templates, see `How to test` for a detailed explanation

A bit about the design:
- Just instructors (course staff) can manage grades
- A student can submit their assignment if and only if: the due date has not passed (configured in the courses' subsection),  the student has not already submitted an assignment for that block, and the student has not already received a score. If one of these conditions is not met, then the buttons `Save assignment` or `Submit` should not appear
- A student can save their assignment any number of times

### How to test
1. Install this xblock's version in your dev environment
2. Create a new course
3. Create a new `Section > Subsection > Unit > Component: Mindmap`
4. You can use these configurations:
5. Add a simple grading configuration to your course:

Your course > Settings > Grading
<img src="https://github.com/eduNEXT/xblock-mindmap/assets/64440265/fdae193a-1cfd-43ce-8a69-4842338641ba)" width="500" height="400" />

Your course > Edit subsection
<img src="https://github.com/eduNEXT/xblock-mindmap/assets/64440265/9d620abc-a373-4117-b82f-8d9eb9ef82a5" width="300" height="200" />

#### Test cases
As a student
- Save your mindmap how many times you want
- Submit your mindmap, reload the page. After reloading the buttons `Save` or `Submit` should not be available

[Screencast from 18-08-23 17:17:22.webm](https://github.com/eduNEXT/xblock-mindmap/assets/64440265/a45dc612-bbc1-4f62-b0fa-cf4fd97c632b)

Then, as an instructor:
- Go to your components, and you'll see a few buttons and inputs:
<img src="https://github.com/eduNEXT/xblock-mindmap/assets/64440265/9be84573-f92c-4db5-8440-8d1d684be2fc" width="300" height="200" />

- Press `Get submissions` to get all the submissions for the current block, then check the dev console:
![image](https://github.com/eduNEXT/xblock-mindmap/assets/64440265/ea108d12-39bd-47e0-97a4-030bc28a5293)

- Enter a grade for the student that submitted the mindmap, by pressing `Submit` with the submission ID and the grade into the inputs:
![image](https://github.com/eduNEXT/xblock-mindmap/assets/64440265/e5eae8b8-da6f-421d-b4f9-46e880c06bbc)

As a student:
- Reload your component, you should see the grade the instructor gave you
![image](https://github.com/eduNEXT/xblock-mindmap/assets/64440265/f4545b9f-e086-4a44-ae17-a2c5db534a2d)

Again, as an instructor:
- Reset the score, by pressing `Remove grade` after filling the student ID you copied from the dev console:
![image](https://github.com/eduNEXT/xblock-mindmap/assets/64440265/1c84a1c5-2593-4597-9309-01c0524de4dc)

As a student:
- You shouldn't see the score after reloading the page.

Now, from the components' staff debug, delete the student state:
![image](https://github.com/eduNEXT/xblock-mindmap/assets/64440265/82f7e87b-579c-45a7-8193-07accc545c73)

Then, as a student, reload the component again (you have no data saved since we deleted it in our previous step).

Now, configure a due date way back in the past for the subsection:
![image](https://github.com/eduNEXT/xblock-mindmap/assets/64440265/2bcd6b49-a69e-4210-b41f-11721c620ae2)

As a student, enter the component:
![image](https://github.com/eduNEXT/xblock-mindmap/assets/64440265/af7a38a5-e0a1-477d-a3a4-41cddfbf9e19)

You'll see that you can't submit the mindmap because the due date passed. Now, as an instructor, add an individual date extension:
Instructor dashboard > Extensions
![image](https://github.com/eduNEXT/xblock-mindmap/assets/64440265/5e979e39-cbc9-490b-8844-85dfb9f5e918)

Then, as a student, submit a new mindmap.

### Questions not resolved yet
- Can a student re-submit their assignment after an instructor resets their score?
